pkg/config/apiserver.go

```diff
@@ -0,0 +1,210 @@
+	ws.Path(a.rootPath)
+
+	ws.Route(ws.
+		GET("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.getRules).
+		Doc("Gets rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(pb.ServiceConfig{}))
+
+	ws.Route(ws.
+		PUT("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.putRules).
+		Doc("Replaces rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(rpc.Status{}))
+	c.Add(ws)
+}
+
+// NewAPI creates a new API server
+func NewAPI(version string, port int, eval expr.Validator, aspectFinder AspectValidatorFinder,
+	builderFinder BuilderValidatorFinder, findAspects AdapterToAspectMapper, store KVStore) *API {
+	c := restful.NewContainer()
+	a := &API{
+		version:  version,
+		rootPath: fmt.Sprintf("/api/%s", version),
+		store:    store,
+		readBody: ioutil.ReadAll,
+		validate: func(cfg map[string]string) (rt *Validated, ce *adapter.ConfigErrors) {
+			v := newValidator(aspectFinder, builderFinder, findAspects, true, eval)
+			return v.validate(cfg)
+		},
+	}
+	a.register(c)
+	a.Server = &http.Server{Addr: ":" + strconv.Itoa(port), Handler: c}
+	a.handler = c
+	return a
+}
+
+// Run calls listen and serve on the API server
+func (a *API) Run() {
+	glog.Infof("Starting Config API Server at %v", a.Server.Addr)
+	glog.Warning(a.Server.ListenAndServe())
+}
+
+// getRules returns the entire service config document for the scope and subject
+// "/scopes/{scope}/subjects/{subject}/rules"
+func (a *API) getRules(req *restful.Request, resp *restful.Response) {
+	funcPath := req.Request.URL.Path[len(a.rootPath):]
+	val, index, found := a.store.Get(funcPath)
+	if !found {
+		writeResponse(http.StatusNotFound, fmt.Sprintf("no rules for %s\n", funcPath), resp)
+		return
+	}
+
+	// TODO send index back to the client
+	_ = index
+	resp.AddHeader("Content-Type", "application/yaml")
+	write(val, resp)
+}
+
+// putRules replaces the entire service config document for the scope and subject
+// "/scopes/{scope}/subjects/{subject}/rules"
+func (a *API) putRules(req *restful.Request, resp *restful.Response) {
+	funcPath := req.Request.URL.Path[len(a.rootPath):]
+
+	// TODO optimize only read descriptors and adapters
+	data, _, index, err := readdb(a.store, "/")
+	if err != nil {
+		writeResponse(http.StatusInternalServerError, err.Error(), resp)
+		return
+	}
+	// TODO send index back to the client
+	_ = index
+
+	var bval []byte
+	bval, err = a.readBody(req.Request.Body)
+	if err != nil {
+		writeResponse(http.StatusInternalServerError, err.Error(), resp)
+		return
+	}
+	val := string(bval)
+	data[funcPath] = val
+
+	_, cerr := a.validate(data)
+	if cerr != nil {
+		glog.Warning(cerr.Error())
+		glog.Warning(val)
+		writeResponse(http.StatusPreconditionFailed, cerr.Error(), resp)
```

Comment:
is this a precondition failure or an invalid request?  It looks like an invalid request.

Also, can we add a bit of user friendliness to the error message here?  Something like: "Request failed validation. Errors: %v", cerr.Error()" maybe?

Guidelines here: https://cloud.google.com/apis/design/errors.